resolver/dns/dns_resolver.go

```diff
@@ -0,0 +1,348 @@
+		d.mu.Lock()
+		d.t.Reset(d.b.freq)
+		d.mu.Unlock()
+		d.cc.NewAddress(result)
+		d.cc.NewServiceConfig(string(sc))
+	}
+}
+
+// ipResolver watches for the name resolution update for an IP address.
+type ipResolver struct {
+	cc         resolver.ClientConn
+	updateChan chan resolver.Address
+}
+
+// ResolveNow is a no-op. For IP address, the resolution is itself, thus
+// re-resolution is unncessary.
+func (i *ipResolver) ResolveNow(opt resolver.ResolveNowOption) {}
+
+// Close closes the ipResolver.
+func (i *ipResolver) Close() {
+	close(i.updateChan)
+}
+
+// dnsResolver watches for the name resolution update for a non-IP target.
+type dnsResolver struct {
+	b      *dnsBuilder
+	host   string
+	port   string
+	ctx    context.Context
+	cancel context.CancelFunc
+	cc     resolver.ClientConn
+	// rn channel is used by ResolveNow() to force an immediate resolution of the target.
+	rn chan struct{}
+	mu sync.Mutex
+	t  *time.Timer
+}
+
+// ResolveNow invoke an immediate resolution of the target that this dnsResolver watches.
+func (d *dnsResolver) ResolveNow(opt resolver.ResolveNowOption) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.t.Stop()
+	d.rn <- struct{}{}
+	d.t.Reset(d.b.freq)
+}
+
+// Close closes the dnsResolver.
+func (d *dnsResolver) Close() {
+	d.cancel()
+}
+
+func (d *dnsResolver) lookupSRV() []resolver.Address {
+	var newAddrs []resolver.Address
+	_, srvs, err := lookupSRV(d.ctx, "grpclb", "tcp", d.host)
+	if err != nil {
+		grpclog.Infof("grpc: failed dns SRV record lookup due to %v.\n", err)
+		return nil
+	}
+	for _, s := range srvs {
+		lbAddrs, err := lookupHost(d.ctx, s.Target)
+		if err != nil {
+			grpclog.Warningf("grpc: failed load banlacer address dns lookup due to %v.\n", err)
+			continue
+		}
+		for _, a := range lbAddrs {
+			a, ok := formatIP(a)
+			if !ok {
+				grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
+				continue
+			}
+			addr := a + ":" + strconv.Itoa(int(s.Port))
+			newAddrs = append(newAddrs, resolver.Address{Addr: addr, Type: resolver.GRPCLB, ServerName: s.Target})
+		}
+	}
+	return newAddrs
+}
+
+func (d *dnsResolver) lookupTXT() []byte {
+	ss, err := lookupTXT(d.ctx, d.host)
+	if err != nil {
+		grpclog.Warningf("grpc: failed dns TXT record lookup due to %v.\n", err)
+		return nil
+	}
+	var res string
+	for _, s := range ss {
+		res += s
+	}
+	return []byte(res)
+}
+
+func (d *dnsResolver) lookupHost() []resolver.Address {
+	var newAddrs []resolver.Address
+	addrs, err := lookupHost(d.ctx, d.host)
+	if err != nil {
+		grpclog.Warningf("grpc: failed dns A record lookup due to %v.\n", err)
+		return nil
+	}
+	for _, a := range addrs {
+		a, ok := formatIP(a)
+		if !ok {
+			grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
+			continue
+		}
+		addr := a + ":" + d.port
+		newAddrs = append(newAddrs, resolver.Address{Addr: addr})
+	}
+	return newAddrs
+}
+
+func (d *dnsResolver) lookup() ([]resolver.Address, []byte) {
+	newAddrs := d.lookupSRV()
+	if newAddrs == nil {
```

Comment:
What if the result contains both balancer addresses and backend addresses?
There's a fallback mechanism that use backends directly if cannot connect to the remote balancers.